providers/unleash/src/main/java/dev/openfeature/contrib/providers/unleash/UnleashProvider.java

```diff
@@ -0,0 +1,171 @@
+    private static final String NAME = "Unleash Provider";
+    public static final String NOT_IMPLEMENTED =
+        "Not implemented - provider does not support this type. Only boolean is supported.";
+
+    public static final String PROVIDER_NOT_YET_INITIALIZED = "provider not yet initialized";
+    public static final String UNKNOWN_ERROR = "unknown error";
+
+    @Getter(AccessLevel.PROTECTED)
+    private UnleashOptions unleashOptions;
+
+    @Setter(AccessLevel.PROTECTED)
+    @Getter
+    private Unleash unleash;
+
+    @Setter(AccessLevel.PROTECTED)
+    @Getter
+    private ProviderState state = ProviderState.NOT_READY;
+
+    private AtomicBoolean isInitialized = new AtomicBoolean(false);
+
+    /**
+     * Constructor.
+     * @param unleashOptions UnleashOptions
+     */
+    public UnleashProvider(UnleashOptions unleashOptions) {
+        this.unleashOptions = unleashOptions;
+    }
+
+    /**
+     * Initialize the provider.
+     * @param evaluationContext evaluation context
+     * @throws Exception on error
+     */
+    @Override
+    public void initialize(EvaluationContext evaluationContext) throws Exception {
+        boolean initialized = isInitialized.getAndSet(true);
+        if (initialized) {
+            throw new GeneralError("already initialized");
+        }
+        super.initialize(evaluationContext);
+        UnleashSubscriberWrapper unleashSubscriberWrapper = new UnleashSubscriberWrapper(
+            unleashOptions.getUnleashConfigBuilder().build().getSubscriber(), this);
+        unleashOptions.getUnleashConfigBuilder().subscriber(unleashSubscriberWrapper);
+        UnleashConfig unleashConfig = unleashOptions.getUnleashConfigBuilder().build();
+        unleash = new DefaultUnleash(unleashConfig);
+
+        // else, state will be changed via UnleashSubscriberWrapper events
+        if (unleashConfig.isSynchronousFetchOnInitialisation()) {
+            state = ProviderState.READY;
+        } else {
+            log.info("ready state will be changed via UnleashSubscriberWrapper events");
+        }
+
+        log.info("finished initializing provider, state: {}", state);
+    }
+
+    @Override
+    public Metadata getMetadata() {
+        return () -> NAME;
+    }
+
+    @Override
+    public void emitProviderReady(ProviderEventDetails details) {
+        super.emitProviderReady(details);
+        state = ProviderState.READY;
+    }
+
+    @Override
+    public void emitProviderError(ProviderEventDetails details) {
+        super.emitProviderError(details);
+        state = ProviderState.ERROR;
+    }
+
+    @Override
+    public ProviderEvaluation<Boolean> getBooleanEvaluation(String key, Boolean defaultValue, EvaluationContext ctx) {
+        if (!ProviderState.READY.equals(state)) {
+            if (ProviderState.NOT_READY.equals(state)) {
+                throw new ProviderNotReadyError(PROVIDER_NOT_YET_INITIALIZED);
+            }
+            throw new GeneralError(UNKNOWN_ERROR);
+        }
```

Comment:
It is an anti-pattern for Unleash SDKs to throw during flag evaluations. It should be safe to check the state of a flag, regardless of the SDK being ready. It leads to a bad coding practice where the user will need to catch runtime exception all over to avoid their application to crash. In most cases a user can rely on default-values and local backups until the SDK has synchronized. If you need your application to not handle load before this has happend you can block on the initialization phase or use the ready event to not make the application ready and added to the LB configuration. 

I am not sure if it is idiomatic to the Open Feature API to throw here, but then it should be taken care of the Core Logic of Open Feature, and not in the wrapper.

Reply:
To clarify, it is taken care of in the Core Logic of Open Feature by [returning a default value](https://github.com/open-feature/java-sdk/blob/main/src/main/java/dev/openfeature/sdk/OpenFeatureClient.java#L150) in this case when evaluating via the client, not expecting the user to catch runtime exception.
It is verified in [provider test](https://github.com/open-feature/java-sdk-contrib/pull/424/files#diff-4ac6957a0b62e15f9248ed31569e8a32d9a7e86a15486f69015b1a40d84ca1dcR180) and [core spec test](url).

\* @toddbaert this is raised by several developers when implementing a provider, if you have some idea how to make it more clear that throwing exception leading to returning default value by the framework, can be taken separately from this PR.

Reply:
If this is important to Open Feature this should live in the core logic and not in the provider. Unleash does not work like this, and this logic changes the exceptions on how Unleash works.

Reply:
I don't think it is changing the logic, please have a look at this [core spec test](https://github.com/open-feature/java-sdk/blob/68f97c6c08a1223ad49e5358cadc3dab66a1d9ec/src/test/java/dev/openfeature/sdk/FlagEvaluationSpecTest.java#L101):

```
assertEquals(ErrorCode.PROVIDER_NOT_READY, client.getBooleanDetails("return_error_when_not_initialized", false).getErrorCode());
```

The user expected to evaluate a flag via the client, not directly via the provider.
It means that if user evaluates a flag via the client, he gets a response with default value and a matching error code, and not an exception.

Reply:
Yes, but I still believe this is cluttering the provider code and open for unnecessary bugs. I strongly recommend that this logic belongs to Open Feature Core and not the provider code.

Reply:
Ok, I hear you. Created core SDK [PR](https://github.com/open-feature/java-sdk/pull/603) for it, then can be removed from this provider.

Reply:
@ivarconr I think you are misunderstanding a point here. An openfeature provider _is not an SDK_. It is an _adapter_ that allows the OpenFeature SDK to talk to a provider certain backend (Unleash in this case). When an OpenFeature _provider_ throws like this _it IS caught by the SDK_ and the user never gets that error.

The error is only recorded internally for debugging.

Reply:
> provider throws like this it IS caught by the SDK and the user never gets that error. This is basically just an example of the "throw early catch late" philosophy. This is the mechanism by which a provider communicates an error to the SDK.

I am really glad Open Feature catches this, and not surface this to the developers. Generally I would be careful on using **exception based control flow**. This tend to scream at me, as I have seen how they play out in large code bases over time. 

> This is the mechanism by which a provider communicates an error to the SDK. The error is only recorded internally for debugging and integration use-cases.

This error is **not theoretically possible** in the Unleash Java SDK, after it has been instantiated. The Unleash Java SDK guarantees that you will get a response from the "isEnabled" call and the "getVariant" call. 

Its up to the developer to decide if he/she want to wait until synchronized (and/or throw) during the initialization phase of Unleash. This is controlled via the Unleash configuration. 

> You could hide all these errors in your provider, but then all sorts of integrations (like the OTel one mentioned above) wouldn't work.

I am not talking about hiding errors, but the implemented error logic here is not relevant for the Unleash Java SDK. 

I also think it needs to be in the concern of OpenFeature core logic to check whether the Provider has been initialized or not as the control logic for that happens outside of the provider code.

Reply:
That's understandable.

I was making general comments on the behavior of the system when providers throw. I agree that we don't want exception based control flow, generally. I was just responding to your general comments about exceptions reaching users, not this particular exception.

Reply:
I prefer to leave the check, as it is possible on incorrect usage:

```
UnleashProvider unleashProvider = new UnleashProvider(unleashOptions);
unleashProvider.getBooleanEvaluation... // -> this should get to the NOT_READY error, as unleash not initialized yet.
OpenFeatureAPI.getInstance().setProviderAndWait("unleash", unleashProvider);
```

Reply:
Ya, that would be odd API usage (not using the OpenFeature evaluation API) but I guess you're right.

Reply:
you make a good point @liran2000. This should not be allowed, and might be caused by my confusion of not fully understanding the Open Feature API. In my understanding using the  `unleashProvider` directly would be an anti-pattern and not expected, or even allowed.